felix/wireguard/bootstrap.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2020-2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"github.com/projectcalico/calico/felix/netlinkshim"
+	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	cerrors "github.com/projectcalico/calico/libcalico-go/lib/errors"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+// BootstrapHostConnectivity forces WireGuard peers with hostencryption enabled to communicate with this node unencrypted.
+// This ensures connectivity in scenarios where we have lost our WireGuard config, but will be sent WireGuard traffic
+// e.g. after a node restart, during felix startup, when we need to fetch config from Typha (calico/issues/5125)
+func BootstrapHostConnectivity(wgDeviceName string, nodeName string, getWireguardHandle func() (netlinkshim.Wireguard, error), calicoClient clientv3.Interface) error {
+	var storedPublicKey string
+	var kernelPublicKey string
+
+	wg, err := getWireguardHandle()
+	if err != nil {
+		log.Debug("Couldn't acquire WireGuard handle, treating pulic key as unset")
+	} else {
+		kernelPublicKey = getPublicKey(wgDeviceName, wg).String()
+		defer wg.Close()
+	}
+
+	// make a few attempts to read our publickey from the datastore, compare, and update if required
+	maxRetries := 3
+	for r := 0; r < maxRetries; r++ {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		thisNode, err := calicoClient.Nodes().Get(ctx, nodeName, options.GetOptions{})
+		cancel()
+		if err != nil {
+			log.WithError(err).Debug("Couldn't fetch node config from datastore, retrying")
+			continue
+		}
+
+		storedPublicKey = thisNode.Status.WireguardPublicKey
+
+		// if there is any config mismatch, wipe the datastore's publickey (forces peers to send unencrypted traffic)
+		if storedPublicKey != kernelPublicKey {
```

Comment:
Just had a thought: if you have a key in the dataplane but it' doesn't match the datastore then you might deadlock like this:
* Remove key from Node; ok, other nodes will send you unencrypted traffic.
* But... you've still got a key in the dataplane so your node might drop _their_ traffic for being unencrypted. (i.e. The opposite of your scenario.)

Does anything prevent that?  Maybe we need to disable wireguard if the keys don't match.

Reply:
Here's an even gnarlier one:

* keys _do_ match (so you skip the cleanup)
* _while felix was down_, the cluster was scaled up, adding new nodes that pre-restart felix didn't know about
* all the typha's are now running on the new nodes

The new nodes see your public key and try to send encrypted traffic but you don't have the peers/allowed IPs configured for them yet so the traffic is dropped.

Reply:
on the former, will unencrypted traffic be dropped by a receiving node if the route back out should be encrypted? I suppose even if it doesnt get dropped there, the receiving node will attempt to send an ACK back over WG which would then be dropped :thinking:

Reply:
I think in the case where our dataplane is *programmed*, but mismatching the datastore, disabling wireguard (or at-least, hostencryption) might be the only option. In that scenario it'd probably make more sense to look at enabling the whole wireguard_mgr again, so that it can cleanup all its dataplane config?

Either than or, potentially, find a way to limit typha discovery to typha's that Felix previously knew about?